onnxscript/values.py

```diff
@@ -183,77 +183,73 @@ def fun(*args, **kwargs):
-            return self._libcall(**kwargs)
-        if isinstance(args[0], tensor.Tensor):
-            return self._libcall(*args, **kwargs)
-        return self._usercall(*args, **kwargs)
-
-    def _usercall(self, *args, **kwargs):
-        """Eager mode"""
-        new_args = []
-        for i, a in enumerate(args):
-            if isinstance(a, np.ndarray):
-                new_args.append(tensor.Tensor(a))
-            elif isinstance(a, (bool, int, float)):
-                new_args.append(tensor.Tensor(np.array(a)))
-            else:
-                raise TypeError(f"Unexpected input type {type(a)} for an input {i}.")
-        res = self.function(*new_args, **kwargs)
-        if isinstance(res, np.ndarray):
-            return res
-        if isinstance(res, tensor.Tensor):
-            return res.value
-        if isinstance(res, (list, tuple)):
-            unwrapped = []
-            for i, r in enumerate(res):
-                if isinstance(r, np.ndarray):
-                    unwrapped.append(r)
-                elif isinstance(r, tensor.Tensor):
-                    unwrapped.append(r.value)
-                else:
-                    raise TypeError(
-                        f"Unexpected output type {type(r)} for an output {i} "
-                        f"in function {self.function!r}."
-                    )
-            if isinstance(res, tuple):
-                return tuple(unwrapped)
-            return unwrapped
-        raise TypeError(f"Unexpected output type {type(res)} in function {self.function!r}.")
-
-    def _libcall(self, *args, **kwargs):
-        """This method must be called when a function decoracted with `script`
-        calls another one decorated with `script`.
-        """
-        new_args = []
-        for i, a in enumerate(args):
-            if isinstance(a, tensor.Tensor):
-                new_args.append(a)
-            elif isinstance(a, bool):
-                # TODO: default values for function parameters
-                # are not properly handled yet. This section
-                # should disappear.
-                new_args.append(tensor.Tensor(np.array(a)))
-            else:
-                raise TypeError(f"Unexpected input type {type(a)} for an input {i}.")
-        res = self.function(*new_args, **kwargs)
-        if isinstance(res, tensor.Tensor):
-            return res
-        if isinstance(res, tuple):
-            unwrapped = []
-            for i, r in enumerate(res):
-                if isinstance(r, tensor.Tensor):
-                    unwrapped.append(r)
-                else:
-                    raise TypeError(
-                        f"Unexpected output type {type(r)} for an output {i} "
-                        f"in function {self.function!r}."
-                    )
-            return tuple(unwrapped)
-        raise TypeError(f"Unexpected output type {type(res)} in function {self.function!r}.")
+        new_args, has_array = self.wrap(args)
```

Comment:
Just for my learning, way is input with nparray treated differently?

Reply:
First, eager-mode execution uses Tensor(...) as a wrapper around numpy arrays to define overloaded operators like +. So, if the user input is a nparray, it is automatically wrapped used a Tensor object.

Second, for user convenience, we unwrap such a Tensor wrapper when returning values to the user at the top-level.

Reply:
I see thanks! I think it'd be very helpful to include this in a comment too